io/cmd.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	cmd := config.GetCmd()
+	cmdReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+	defer cmdReader.Close()
+	scanner := bufio.NewScanner(cmdReader)
+	cmdReaderStderr, err := cmd.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+	defer cmdReaderStderr.Close()
+	scannerStderr := bufio.NewScanner(cmdReaderStderr)
+	err = cmd.Start()
+	if err != nil {
+		return "", err
+	}
+	errChan := make(chan error)
+	var stdoutOutput string
+	wg.Add(1)
+	go func() {
+		for scanner.Scan() {
+			line := scanner.Text()
+			for _, regExp := range regExpStruct {
+				matched := regExp.RegExp.Match([]byte(line))
+				if matched {
+					regExp.matchedResults = regExp.RegExp.FindStringSubmatch(line)
+					regExp.line = line
+					line, err = regExp.ExecFunc()
+					if err != nil {
+						errChan <- err
+					}
+				}
+			}
+			fmt.Println(line)
+			stdoutOutput += line + "\n"
+		}
+		wg.Done()
+	}()
+	wg.Add(1)
+	go func() {
+		for scannerStderr.Scan() {
+			line := scannerStderr.Text()
+			var scannerError error
+			for _, regExp := range regExpStruct {
+				matched := regExp.RegExp.Match([]byte(line))
+				if matched {
+					regExp.matchedResults = regExp.RegExp.FindStringSubmatch(line)
+					regExp.line = line
+					line, scannerError = regExp.ExecFunc()
+					if scannerError != nil {
+						errChan <- scannerError
+						break
+					}
+				}
+			}
+			fmt.Fprintf(os.Stderr, line+"\n")
+			if scannerError != nil {
+				break
+			}
+		}
+		wg.Done()
+	}()
+
+	go func() {
+		wg.Wait()
+		close(errChan)
+	}()
+
+	for err := range errChan {
+		return stdoutOutput, err
+	}
+	return stdoutOutput, nil
+}
+
+type CmdConfig interface {
+	GetCmd() *exec.Cmd
+	GetEnv() map[string]string
+	GetStdWriter() io.WriteCloser
+	GetErrWriter() io.WriteCloser
+}
+
+func GetRegExp(regex string) (*regexp.Regexp, error) {
+	regExp, err := regexp.Compile(regex)
+	if err != nil {
+		return nil, err
+	}
+
+	return regExp, nil
+}
+
+// Mask the credentials information from the line. The credentials are build as user:password
+// For example: http://user:password@127.0.0.1:8081/artifactory/path/to/repo
+func (reg *CmdOutputPattern) MaskCredentials() (string, error) {
+	splittedResult := strings.Split(reg.matchedResults[0], "//")
+	return strings.Replace(reg.line, reg.matchedResults[0], splittedResult[0]+"//***.***@", 1), nil
+}
+
+func (reg *CmdOutputPattern) Error() (string, error) {
+	fmt.Fprintf(os.Stderr, reg.line)
+	if len(reg.matchedResults) > 1 {
+		return "", errors.New(reg.ErrorMessage + strings.TrimSpace(reg.matchedResults[1]))
+	}
+	return "", errors.New(reg.ErrorMessage)
+}
+
+// RegExp - The regexp that the line will be searched upon.
+// matchedResults - The slice result that was found by the regex
+// line - The output line from the external process
+// ErrorMessage - Error message or part of the error that will be returned
+// ExecFunc - The function to execute
+type CmdOutputPattern struct {
+	RegExp         *regexp.Regexp
+	matchedResults []string
```

Comment:
I'm not sure about this, but shouldn't *matchedResults* and *line* be public?

Reply:
No, those are being used only in this class.